### Problem Statement

Compute the value of $e^{0.5}$ using the Taylor series expansion of the exponential function. Determine the smallest number of terms $n$ such that the approximation error is less than $10^{-3}$.

$$
e^{x} = \sum_{k=0}^{\infty} \frac{x^{k}}{k!} = 1 + x + \frac{x^{2}}{2!} + \frac{x^{3}}{3!} + \dots
$$

In this problem, we approximate $e^{0.5}$ by truncating the series after $n$ terms:

$$
e^{0.5} \approx \sum_{k=0}^{n} \frac{(0.5)^{k}}{k!}
$$


In [34]:
import numpy as np

### Factorial Function

In order to build the Taylor series approximation, we first need a function that computes the factorial of a non‑negative integer (n):



In [ ]:
def factorial(n):
    if n == 0 : 
        return 1
    a = 1
    for i in range(1, n+1):
        p = a * i
        a = p
    
    return p


### Taylor Series Approximation Function

To approximate the value of the exponential function $e^x$, we use the truncated Taylor series:

$$
e^{x} \approx \sum_{k=0}^{n} \frac{x^{k}}{k!}
$$

### Description

This function computes the Taylor series approximation of $e^{x}$ using the first $(n)$ terms of the expansion. Each term of the form

$$
\frac{x^i}{i!}
$$

is calculated and added to a cumulative sum, producing an approximation of the exponential function.

### Purpose in This Project

This function plays a central role in the numerical experiment:

- It provides an approximate value of $e^{0.5}$ for a given number of terms $n$.
- By increasing $n$, the approximation becomes more accurate.
- The computed approximation can be compared with the true value $e^{0.5}$ obtained from NumPy.
- Using this function, we can analyze approximation errors and determine the smallest value of $n$ for which the error becomes smaller than $10^{-3}$.


In [35]:
def taylor_exp(x, n):
    s = 0
    for i in range(n+1):
        s += (np.pow(x, i)) / factorial(i)
    
    return s


### True Error Functions

To evaluate the quality of our Taylor series approximation, we compute two kinds of errors:

1. **True Absolute Error**
2. **True Relative Error**

These measures tell us how close the approximation is to the actual value of $e^{x}$.

---

### 1. True Absolute Error

The true absolute error compares the exact value of the function with the approximation:

$$
E_{\text{true}} = \left| e^{x} - \text{TaylorApprox}(x, n) \right|
$$

**Purpose:**  
This shows how far the approximation is from the real value in absolute terms.

---

### 2. True Relative Error

The true relative error normalizes the absolute error by dividing it by the exact value:

$$
E_{\text{true, rel}} =
\frac{\left| e^{x} - \text{TaylorApprox}(x, n) \right|}{\left| e^{x} \right|}
$$

**Purpose:**  
This shows the error as a fraction of the true value, giving a scale‑independent measure of accuracy.


In [36]:
def error_true(x, n):
    error_true = np.abs(np.exp(x) - taylor_exp(x, n))
    return error_true

In [37]:
def error_relative_true(x, n):
    error_relative_true = error_true(x, n) / np.abs(np.exp(x))
    return error_relative_true

### Finding Minimum Terms ($n$) Based on Truncation Error

The goal is to find the minimum number of terms, $n$, for the Taylor expansion of $e^{x}$ such that the approximation error is less than a specified tolerance, $\epsilon$.

For the exponential function, the true error is often approximated by the magnitude of the first neglected term (the $(n+1)$-th term):

$$
E_n(x) \approx \left| \frac{x^{n+1}}{(n+1)!} \right|
$$

We require this estimated error to be less than a tolerance $\epsilon$ (where $\epsilon = (1/2) * 10^{-3}$ in this problem):

$$
\left| \frac{x^{n+1}}{(n+1)!} \right| < \epsilon
$$

The value of $i$ at which this condition is met corresponds to $n+1$ in the formula above. Therefore, the required number of terms $n$ is found by subtracting 1 from this index $i$:

$$
n = i_{\text{stop}} - 1
$$


In [41]:
number = 0.5
accuracy = 10 ** (-3)


def find_the_number_of_sentences(x):
    i = 0
    while True:
        e_a = (np.pow(x, i)) / factorial(i)
        if e_a <= (1/2) * accuracy:
            n = i
            break
        else:
            i += 1
    
    return i

n = find_the_number_of_sentences(number) - 1


In [45]:
print(f"Table for computing the value of e^{number} up to the n term to achieve the desired accuracy.\n")

print(f"{'n':>5} {'y_app':>12} {'error_true':>15} {'error_approx':>15}")
print('-' * 50)

for i in range(0, n+2):
    y_app = taylor_exp(number, i)
    err_tr = error_true(number, i)
    err_rel_tr = error_relative_true(number, i)

    print(f"{i:5d} {y_app:12.5f} {err_tr:15.5f} {err_rel_tr:15.5f}")


Table for computing the value of e^0.5 up to the n term to achieve the desired accuracy.

    n        y_app      error_true    error_approx
--------------------------------------------------
    0      1.00000         0.64872         0.39347
    1      1.50000         0.14872         0.09020
    2      1.62500         0.02372         0.01439
    3      1.64583         0.00289         0.00175
    4      1.64844         0.00028         0.00017
    5      1.64870         0.00002         0.00001


Based on the output table, it is observed that as $n$ increases, the error gradually decreases.  
At $n = 4$, the error becomes sufficiently small to meet the desired accuracy, so this value is selected as the appropriate $n$.

The Taylor series approximation of $e^{0.5}$ up to the 4th term is:

$$
e^{0.5} = 1 + 0.5 + \frac{0.5^2}{2} + \frac{0.5^3}{3!} + \frac{0.5^4}{4!} = 1.64844
$$

Hence, the approximate value of the function for $n = 4$ is **1.64844**.
